### PR TITLE
SurfaceImage and maskers

### PR DESCRIPTION
**Pull request by jeromedockes**
this continues #3799 which was merged by mistake

### CODE REVIEW

**Review by ymzayek — COMMENTED** [review]
LGTM! I can work on unit tests and push them to this PR

File: examples/08_experimental/surface_image_and_maskers.py
Comment (ymzayek) [inline examples/08_experimental/surface_image_and_maskers.py:94]: Eventually what would be the suggested solution(s) here? Adjusting these functions to accept the surface masker or writing new separate functions to check the surface masker if such object is given to the decoder?

File: nilearn/experimental/surface/_datasets.py
Comment (jeromedockes) [reply nilearn/experimental/surface/_datasets.py:17]: I don't like having the function return different things depending on the `mesh_type` argument -- but if you want to add mesh_type (without a default) that's fine, and it's not a big deal to need to call the function a couple of times if we need 2 meshes

File: examples/08_experimental/surface_image_and_maskers.py
Comment (jeromedockes) [reply examples/08_experimental/surface_image_and_maskers.py:94]: I think `_check_embedded_nifti_masker` would have to be adapted (and maybe renamed) so that it can also work for surface maskers. for example if the passed estimator's mask_img is a surface image a surface masker would be returned.

it would have been better if the decoder objects were passed a masker rather than all the masker's parameters but I guess it is too late to change that

File: nilearn/experimental/surface/_datasets.py
Comment (ymzayek) [reply nilearn/experimental/surface/_datasets.py:17]: Ok fair point. Let's leave it like this for now then

**Review by RaphaelMeudec — COMMENTED** [review]
The API is looking good!

I just have a question regarding the `Mesh | InMemoryMesh | FileMesh`. Why is there a need to have a separation between those? Couldn't there be a unique `Mesh(coordinates: Nullable[np.ndarray],  faces: Nullable[np.ndarray])` with a `from_disk` or `load_from_disk` method? I feel it would be easier for the user to have a single class to refer to. Sorry if it's been discussed else where, I couldn't find it in the previous PR.

File: examples/08_experimental/surface_image_and_maskers.py
Comment (jeromedockes) [reply examples/08_experimental/surface_image_and_maskers.py:39]: re this one and other assertions you commented on -- they were rather meant to make the intention clearer for us while discussing the API. you're right they need to be cleaned up or replaced with more useful exceptions

File: nilearn/experimental/surface/_maskers.py
Comment (jeromedockes) [reply nilearn/experimental/surface/_maskers.py:27]: yes I think that's the signature the `fit` method of scikit-learn-compatible estimators are supposed to have, `*_` would work fine too I don't really have a preference

File: nilearn/experimental/surface/_maskers.py
Comment (jeromedockes) [reply nilearn/experimental/surface/_maskers.py:43]: yes this should be replaced with something like the `_check_fitted` method used in other maskers

File: nilearn/experimental/surface/_surface_image.py
Comment (jeromedockes) [reply nilearn/experimental/surface/_surface_image.py:35]: the `__init__` should take care of making it a Path so we can then rely on it being a Path and so we deal with that in the init only instead of everywhere

File: nilearn/experimental/surface/_surface_image.py
Comment (jeromedockes) [reply nilearn/experimental/surface/_surface_image.py:90]: not sure about AttributeError, maybe ValueError instead but otherwise agreed

**Comment by jeromedockes** [comment]
> I just have a question regarding the Mesh | InMemoryMesh | FileMesh. Why is there a need to have a separation between those? Couldn't there be a unique Mesh(coordinates: Nullable[np.ndarray],  faces: Nullable[np.ndarray]) with a from_disk or load_from_disk method? I feel it would be easier for the user to have a single class to refer to. Sorry if it's been discussed else where, I couldn't find it in the previous PR.

they both have to expose a `coordinates() & faces()` interface, but they implement it differently, in one case by storing a file path and in the other case by storing 2 arrays. so that's sort of a classic use case for having 2 different classes

having one class with all those attributes means we always have to check in which configuration we are, and there are possibilities for having incorrect ones (eg all are missing, or they conflict), and having those potentially null attributes is a [classic](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/) source of bugs (when one forgets to handle the `None` case)

I do agree we should have something like `load` method that transforms an on-disk mesh into a in-memory one

File: examples/08_experimental/surface_image_and_maskers.py
Comment (jeromedockes) [reply examples/08_experimental/surface_image_and_maskers.py:39]: actually that one was probably to make mypy happy

**Comment by jeromedockes** [comment]
what name shall we use for keys in the PolyMesh or PolyData (which will typically be something like "left_hemisphere" and "right_hemisphere") when we need to refer them in code (eg to iterate over them) -- something like "part"?

**Review by ymzayek — COMMENTED** [review]
@jeromedockes thanks for adding the fixtures and developing on the tests. I think in its current state this pretty much covers the initial and followup discussions. Maybe a bit more testing is needed but it LGTM. And the failing tests only need the datatype parameter to be added to wherever `GiftiDataArray` is called

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
- [x] add warning when importing that let users know this is WIP / experimental / use-at-your-own-risk ...
- [x] make it in your face obvious in the example that this feature and its API is experimental...

File: nilearn/experimental/surface/_io.py
Comment (ymzayek) [inline nilearn/experimental/surface/_io.py:27]:
```suggestion
        coordinates, intent="NIFTI_INTENT_POINTSET", datatype="float32"
```

**Comment by jeromedockes** [comment]
thanks @ymzayek ! I can add some docstrings to make flake8 happy, and a whatsnew entry.
actually before we merge I wanted to improve a bit the masker for the very frequent use case where we don't want to mask anything by avoiding to store the mask img. I was thinking maybe of having a `SurfaceVectorizer` that just checks shapes and concatenates data but doesn't have a mask_img and doesn't mask anything

**Review by Remi-Gau — APPROVED** [review]
OK it's good with me

**Comment by Remi-Gau** [comment]
Let's wait till CI is done and then merge

**Comment by Remi-Gau** [comment]
I will fix the doc failure in #4014

**Pull request merged**
